core/services/gateway/network/wsconnection.go

```diff
@@ -0,0 +1,103 @@
+package network
+
+import (
+	"errors"
+	"sync/atomic"
+
+	"github.com/gorilla/websocket"
+)
+
+// WSConnectionWrapper is a websocket connection abstraction that supports re-connects.
+// I/O is separated from connection management:
+//   - component doing I/O provides read and write channels that never change
+//   - component managing connections can listen to connection-closed events and call Restart()
+//     to swap the underlying connection object
+//
+// The Wrapper can be used by a server expecting long-lived connections from a given client,
+// as well as a client maintaining such long-lived connection with a given server.
+// This fits the Gateway very well as servers accept connections only from a fixed set of nodes
+// and conversely, nodes only connect to a fixed set of servers (Gateways).
+//
+// The concept of "pumps" is borrowed from https://github.com/smartcontractkit/wsrpc
+type WSConnectionWrapper struct {
+	conn atomic.Pointer[websocket.Conn]
+
+	writeCh    <-chan WriteItem
+	readCh     chan<- ReadItem
+	shutdownCh chan struct{}
+}
+
+type ReadItem struct {
+	MsgType int
+	Data    []byte
+}
+
+type WriteItem struct {
+	MsgType int
+	Data    []byte
+	ErrCh   chan error
```

Comment:
Why do you ask user to put `errCh` into every message? Why not to supply the `errCh` in `NewWSConnectionWrapper` along with read/write Ch? Most likely there will be one `errCh` created by user.

Reply:
That's where I missed the usage. Okay, if the intention is to write concurrently from many goroutines then good.

Reply:
Then I would rather have a `.WriteMessage(msg) error` method on the wrapper.
Otherwise each goroutine needs to manage two channels, properly closing them, etc.

Reply:
OK, turns out a WriteMessage() method is also convenient to use. I'm adding it, thanks!